monai/data/dataset.py

```diff
@@ -321,6 +323,238 @@ def __getitem__(self, index):
+
+        self._start_pos: int = 0
+        self._update_lock: threading.Lock = threading.Lock()
+        self._round: int = 1
+        self._round_done: bool = False
+        self._replace_mgr: threading.Thread = threading.Thread(target=_manage_replacement, args=(self,))
+        self._started: bool = False
+
+        self._compute_data_idx()
+
+    def _compute_data_idx(self):
+        """
+        Update the replacement data position in the total data.
+
+        """
+        for i in range(self._replace_num):
+            pos: int = self._start_pos + self.cache_num + i
+            if pos >= self._total_num:
+                pos -= self._total_num
+            self._replace_data_idx[i] = pos
+
+    def start(self):
+        """
+        Start the background thread to replace training items for every epoch.
+
+        """
+        self._replace_mgr.start()
+        self._started = True
+
+    def _restart(self):
+        """
+        Restart background thread if killed for some reason.
+
+        """
+        self._round = 1
+        self._replace_mgr = threading.Thread(target=_manage_replacement, args=(self,))
+        self.start()
+
+    def _try_update_cache(self):
+        """
+        Update the cache items with new replacement for current epoch.
+
+        """
+        with self._update_lock:
+            if self._round_done:
+                remain_num: int = self.cache_num - self._replace_num
+                for i in range(remain_num):
+                    self._cache[i] = self._cache[i + self._replace_num]
+                for i in range(self._replace_num):
+                    self._cache[remain_num + i] = self._replacements[i]
+
+                self._start_pos += self._replace_num
+                if self._start_pos >= self._total_num:
+                    self._start_pos -= self._total_num
+
+                self._compute_data_idx()
+
+                # ready for next round
+                self._round += 1
+                self._round_done = False
+                return True
+            else:
+                return False
+
+    def update_cache(self):
+        """
+        Update cache items for current epoch, need to call this function before every epoch.
+        If the cache has been shutdown before, need to restart the `_replace_mgr` thread.
+
+        """
+        if not self._replace_mgr.isAlive():
+            self._restart()
+
+        # make sure update is done
+        while True:
+            if self._try_update_cache():
+                break
+            else:
+                time.sleep(0.01)
+
+    def _try_shutdown(self):
+        """
+        Wait for thread lock to shut down the background thread.
+
+        """
+        with self._update_lock:
+            if self._round_done:
+                self._round = 0
+                self._round_done = False
+                return True
+            else:
+                return False
+
+    def shutdown(self):
```

Comment:
The problem with a shutdown method like this is related to the daemon issue, if an exception causes early termination this won't get called so the user has to be vigilant about making sure it is through a try/finally block somewhere.

Reply:
Hi @ericspod ,

Do you mean to use `threading.Thread(target=_manage_replacement, deamon=True, args=(self,))` to fix it?

Thanks.

Reply:
Yes setting `daemon=True` will ensure the thread doesn't hold up program termination, which is safe since it isn't doing anything which must be completed. You can use `threading.Thread(target=self.manage_replacement, daemon=True)` as @wyli mentioned.

Reply:
sure, updated in the latest commit.
Thanks.